third_party/move/move-vm/runtime/src/storage/code_storage.rs

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{loader::Script, ModuleStorage};
+use ambassador::delegatable_trait;
+use move_binary_format::{errors::VMResult, file_format::CompiledScript};
+use sha3::{Digest, Sha3_256};
+use std::sync::Arc;
+
+/// Returns the hash (SHA-3-256) of the script.
+pub fn script_hash(serialized_script: &[u8]) -> [u8; 32] {
+    let mut sha3_256 = Sha3_256::new();
+    sha3_256.update(serialized_script);
+    sha3_256.finalize().into()
+}
+
+/// Represents storage which in addition to modules, also caches scripts. The clients can implement
+/// this trait to ensure that even script dependency is upgraded, the correct script is still
+/// returned. Scripts are cached based on their hash.
+#[delegatable_trait]
+pub trait CodeStorage: ModuleStorage {
```

Comment:
Rename it to script storage? Code seems a bit too generic here.

Reply:
But `ScriptStorage: ModuleStorage` seems weird? Maybe `ModuleAndScriptStorsge` because this is what it is?